text/0074-global-init.md

```diff
@@ -0,0 +1,176 @@
+
+Where the language provides a commonly accepted way to inject SDK
+components, it should be preferred.  The Java SPI (Service Provider
+Interface) supports loading and configuring the SDK before it is first
+used.  This kind of support is the preferred choice in languages with
+common support for automatic dependency injection.
+
+In this case, there is no use-before-configuration question to
+address, as there is in some languages with support for static
+initialization.
+
+### Explicit initializer mechanism
+
+When it is not possible to ensure the SDK is installed and configured
+before the API is first used, initializing the default, global SDK
+instances becomes the user's responsibility.
+
+Methods to set the global instances shall be independent, allowing
+each SDK component to be intialized separately when the process
+starts.  The methods shall be declared in a separate API package,
+e.g., named `global`:
+
+```golang
+package global
+
+import (
+    "go.opentelemetry.io/otel/api/context/propagation"
+    "go.opentelemetry.io/otel/api/metric"
+    "go.opentelemetry.io/otel/api/trace"
+)
+
+// SetMeterProvider initializes the global Meter provider.  May only
+// be called once per process lifetime.  Subsequent calls will panic.
+//
+// Prior to setting the global Meter provider, the default global
+// Meter provider acts as a "forwarding" SDK.  The default global
+// Meter provider will begin forwarding to the installed Meter
+// provider once this is called.
+func SetMeterProvider(metric.Provider) { ... }
+
+// SetPropagators initializes the global Propagators instance.  May only
+// be called once per process lifetime.  Subsequent calls will panic.
+//
+// Prior to setting the global Meter provider, the default global
+// Propagators instance performs pass-through W3C Traceparent and
+// Correlation-Context propagation.
+func SetPropagators(propagation.Propagators) { ... }
+
+// SetPropagators initializes the global Tracer provider.  May only
+// be called once per process lifetime.  Subsequent calls will panic.
+//
+// Prior to setting the global Tracer provider, the default global
+// Tracer provider acts as a "forwarding" SDK.  The default global
+// Tracer provider will begin forwarding to the installed Tracer
+// provider once this is called.
+func SetTraceProvider(trace.Provider) { ... }
+```
+
+#### Requirements
+
+We anticipate third party libraries using the global instances before
+they are installed, and we wish for references obtained through these
+instances to become functional once the corresponding implementation
+is initialized.  The default instances returned by the global getters
+for Tracer provider, Meter provider, and Propagators must "forward" to
+the real SDK implementation once it is installed.
+
+#### Tracer
+
+Tracers obtained through the provider will become functional when the
+user's Tracer SDK is installed as the global instance.
+
+Spans started prior to installing the Tracer SDK will be No-op spans.
```

Comment:
This requires Noop to be part of the API (in some form) so that real tracers can distinguish Noop spans from their own spans. The spec needs to explain this.

Reply:
In Java we 'already' can as we have `DefaultSpan` (previously called `BlankSpan`), as well as `DefaultTracer` and related components. And IIRC other SIGs have something like this as part of the API. But we definitely need to handle them properly (to prevent exceptions all over the place).

Reply:
Why is it a requirement for the tracer to distinguish its own spans?  I'm not aware of any circumstances that would be needed in the present API.

In the accompanying Go PR (which deals in Metrics), I created a new interface for this sort of reason. If the global default creates a `LabelSet` and the real SDK sees it, it must test for conversion to `LabelSetDelegate` and call its `Delegate()` method to get its own `LabelSet`.

This also reminds me -- there are reasons why the **user** wants to know if they're dealing with a No-op span. I'm not sure we've settled the question of how a user can ask this question in the specification. (I generally think of this as a mechanism of last resort, I would prefer to support "lazy" values that are dispatched cheaply when there is not a real span, but the requirement to know whether a span is "real" somehow never goes away completely.)